### PR TITLE
Added certificate serial number column

### PR DESCRIPTION
**Pull request by gabipurcaru**
Adding a new column called "serial" to the certificates table, on Darwin and Windows. This is a hex-encoded string, e.g. `2E9EF8E848F2AAFC37C4A1E5A1B67C6`. 

    osquery> select serial from certificates limit 1;
    +----------+
    | serial   |
    +----------+
    | 2ACEAE9A |
    +----------+

I've tested the change on both Windows 10 and OSX 10.13.3.

### CODE REVIEW

**Comment by muffins** [comment]
ok to test

**Comment by obelisk** [comment]
Can you please fill out the description more?

**Comment by osqueryer** [comment]
:-1: The commit 88041037be4e213a412b433aabd948925d84c723 (Job results: [6876](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6876/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 88041037be4e213a412b433aabd948925d84c723 (Job results: [6884](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6884/)) failed one or more tests (**Linux**).

**Comment by gabipurcaru** [comment]
/usr/local/osquery/bin/ld.lld: error: failed to open osquery_additional_tests: No space left on device

This is the actual CI failure, seems unrelated.

**Comment by gabipurcaru** [comment]
ok to test

File: osquery/tables/system/darwin/keychain_utils.cpp [hunk 1]
```diff
@@ -133,6 +133,12 @@ std::string genSHA1ForCertificate(X509* cert) {
   return "";
 }
 
+std::string genSerialForCertificate(X509* cert) {
+  ASN1_INTEGER* serial = X509_get_serialNumber(cert);
+  BIGNUM* bignumSerial = ASN1_INTEGER_to_BN(serial, NULL);
```
Comment (theopolis) [inline osquery/tables/system/darwin/keychain_utils.cpp]: nit, prefer nullptr over NULL, this results in better static analysis.

File: osquery/tables/system/darwin/keychain_utils.cpp [hunk 2]
```diff
@@ -133,6 +133,12 @@ std::string genSHA1ForCertificate(X509* cert) {
...
+  return BN_bn2hex(bignumSerial);
```
Comment (theopolis) [inline osquery/tables/system/darwin/keychain_utils.cpp]: I think there’s a memory leak here. From the OpenSSL docs it looks like the caller is responsible for the return from both ASN1_Integrr_to_BN and BN_bn2hex, but I just skimmed the docs so I’m not positive.

**Comment by gabipurcaru** [comment]
Thanks for the review, @theopolis, and please excuse my rusty C++. I've updated the code to do null checks and memory deallocation.

File: osquery/tables/system/darwin/keychain_utils.cpp [hunk 3]
```diff
@@ -133,6 +133,23 @@ std::string genSHA1ForCertificate(X509* cert) {
   return "";
 }
 
+std::string genSerialForCertificate(X509* cert) {
+  std::string hex;
+  ASN1_INTEGER* serial = X509_get_serialNumber(cert);
+  BIGNUM* bignumSerial = ASN1_INTEGER_to_BN(serial, nullptr);
+  if (bignumSerial == nullptr) {
+    return hex;
```
Comment (fmanco) [inline osquery/tables/system/darwin/keychain_utils.cpp]: Need to free here.

File: osquery/tables/system/darwin/keychain_utils.cpp [hunk 4]
```diff
@@ -133,6 +133,23 @@ std::string genSHA1ForCertificate(X509* cert) {
...
+  }
+  char* hexBytes = BN_bn2hex(bignumSerial);
+  if (hexBytes == nullptr) {
+    return hex;
```
Comment (fmanco) [inline osquery/tables/system/darwin/keychain_utils.cpp]: And here.

File: osquery/tables/system/darwin/keychain_utils.cpp (hunk 3 above)
Comment (fmanco) [reply osquery/tables/system/darwin/keychain_utils.cpp]: Actually here you don't since serial shouldn't be freed, sorry.

**Pull request merged**